// @jupyterlab/apputils/lib/tokens.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
import { Token } from '@lumino/coreutils';
/**
 * The command palette token.
 */
export const ICommandPalette = new Token('@jupyterlab/apputils:ICommandPalette');
export const IKernelStatusModel = new Token('@jupyterlab/apputils:IKernelStatusModel');
/**
 * The session context dialogs token.
 */
export const ISessionContextDialogs = new Token('@jupyterlab/apputils:ISessionContextDialogs');
/**
 * The theme manager token.
 */
export const IThemeManager = new Token('@jupyterlab/apputils:IThemeManager');
/**
 * The sanitizer token.
 */
export const ISanitizer = new Token('@jupyterlab/apputils:ISanitizer');
/**
 * The main menu token.
 */
export const ISplashScreen = new Token('@jupyterlab/apputils:ISplashScreen');
/**
 * The default window resolver token.
 */
export const IWindowResolver = new Token('@jupyterlab/apputils:IWindowResolver');
/**
 * The toolbar registry token.
 */
export const IToolbarWidgetRegistry = new Token('@jupyterlab/apputils:IToolbarWidgetRegistry');
